xbmc/guilib/GUIVisualisationControl.cpp

```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+  track.genre       = genre.c_str();
+  track.comment     = tag->GetComment().c_str();
+  track.lyrics      = tag->GetLyrics().c_str();
+  track.trackNumber = tag->GetTrackNumber();
+  track.discNumber  = tag->GetDiscNumber();
+  track.duration    = tag->GetDuration();
+  track.year        = tag->GetYear();
+  track.rating      = tag->GetUserrating();
+
+  m_instance->OnAction(VIS_ACTION_UPDATE_TRACK, &track);
+}
+
+bool CGUIVisualisationControl::IsLocked()
+{
+  if (m_instance && m_alreadyStarted)
+    return m_instance->IsLocked();
+
+  return false;
+}
+
+bool CGUIVisualisationControl::HasPresets()
+{
+  if (m_instance && m_alreadyStarted)
+    return m_instance->HasPresets();
+
+  return false;
+}
+
+int CGUIVisualisationControl::GetActivePreset()
+{
+  if (m_instance && m_alreadyStarted)
+    return m_instance->GetActivePreset();
+
+  return -1;
+}
+
+void CGUIVisualisationControl::SetPreset(int idx)
+{
+  if (m_instance && m_alreadyStarted)
+    m_instance->OnAction(VIS_ACTION_LOAD_PRESET, static_cast<void*>(&idx));
+}
+
+std::string CGUIVisualisationControl::GetActivePresetName()
+{
+  if (m_instance && m_alreadyStarted)
+    return m_instance->GetActivePresetName();
+
+  return "";
+}
+
+bool CGUIVisualisationControl::GetPresetList(std::vector<std::string> &vecpresets)
+{
+  if (m_instance && m_alreadyStarted)
+    return m_instance->GetPresetList(vecpresets);
+
+  return false;
+}
+
+bool CGUIVisualisationControl::InitVisualization()
+{
+  ADDON::AddonPtr addonInfo;
+  ADDON::CAddonMgr::GetInstance().GetAddon(CServiceBroker::GetSettings().GetString(CSettings::SETTING_MUSICPLAYER_VISUALISATION), addonInfo, ADDON::ADDON_VIZ, true);
+  if (!addonInfo)
+    return false;
+
+  CServiceBroker::GetActiveAE().RegisterAudioCallback(this);
+
+  g_graphicsContext.CaptureStateBlock();
+
+  float x = g_graphicsContext.ScaleFinalXCoord(GetXPosition(), GetYPosition());
+  float y = g_graphicsContext.ScaleFinalYCoord(GetXPosition(), GetYPosition());
+  float w = g_graphicsContext.ScaleFinalXCoord(GetXPosition() + GetWidth(), GetYPosition() + GetHeight()) - x;
+  float h = g_graphicsContext.ScaleFinalYCoord(GetXPosition() + GetWidth(), GetYPosition() + GetHeight()) - y;
+  if (x < 0)
+    x = 0;
+  if (y < 0)
+    y = 0;
+  if (x + w > g_graphicsContext.GetWidth())
+    w = g_graphicsContext.GetWidth() - x;
+  if (y + h > g_graphicsContext.GetHeight())
+    h = g_graphicsContext.GetHeight() - y;
+
+  m_instance = new ADDON::CVisualization(std::dynamic_pointer_cast<ADDON::CAddonDll>(addonInfo), x, y, w, h);
+  CreateBuffers();
+
+  m_alreadyStarted = false;
+  g_graphicsContext.ApplyStateBlock();
+  return true;
+}
+
+void CGUIVisualisationControl::DeInitVisualization()
+{
+  if (!m_attemptedLoad)
+    return;
+
+  /*
+   * Prevent "UnregisterAudioCallback" in case Kodi is stopped. There becomes
+   * audio engine stopped before gui!
+   */
+  if (!g_application.m_bStop)
+    CServiceBroker::GetActiveAE().UnregisterAudioCallback(this);
```

Comment:
This might race. Make ActiveAE always exist like this: https://github.com/xbmc/xbmc/blob/master/xbmc/ServiceManager.cpp#L40

That way, `UnregisterAudioCallback(this)` will just have no effect after ActiveAE is deinitialized